tests/TestRunner/local_cluster.py

```diff
@@ -0,0 +1,111 @@
+from pathlib import Path
+from argparse import ArgumentParser
+import random
+import string
+import subprocess
+import sys
+
+class LocalCluster:
+    configuration_template = """
+## foundationdb.conf
+##
+## Configuration file for FoundationDB server processes
+## Full documentation is available at
+## https://apple.github.io/foundationdb/configuration.html#the-configuration-file
+
+[fdbmonitor]
+
+[general]
+restart_delay = 10
+## by default, restart_backoff = restart_delay_reset_interval = restart_delay
+# initial_restart_delay = 0
+# restart_backoff = 60
+# restart_delay_reset_interval = 60
+cluster_file = {etcdir}/fdb.cluster
+# delete_envvars =
+# kill_on_configuration_change = true
+
+## Default parameters for individual fdbserver processes
+[fdbserver]
+command = {fdbserver_bin}
+public_address = auto:$ID
+listen_address = public
+datadir = {datadir}
+logdir = {logdir}
+# logsize = 10MiB
+# maxlogssize = 100MiB
+# machine_id =
+# datacenter_id =
+# class =
+# memory = 8GiB
+# storage_memory = 1GiB
+# cache_memory = 2GiB
+# metrics_cluster =
+# metrics_prefix =
+
+## An individual fdbserver process with id 4000
+## Parameters set here override defaults from the [fdbserver] section
+[fdbserver.{server_port}]
+    """
+
+    valid_letters_for_secret = string.ascii_letters + string.digits
+
+    def __init__(self, basedir: str, fdbserver_binary: str, fdbmonitor_binary: str,
+                 fdbcli_binary: str, create_config=True, port=None, ip_address=None):
+        self.basedir = Path(basedir)
+        self.fdbserver_binary = Path(fdbserver_binary)
+        self.fdbmonitor_binary = Path(fdbmonitor_binary)
+        self.fdbcli_binary = Path(fdbcli_binary)
+        for b in (self.fdbserver_binary, self.fdbmonitor_binary, self.fdbcli_binary):
+            assert b.exists(), "{} does not exist".format(b)
+        if not self.basedir.exists():
+            self.basedir.mkdir()
+        self.etc = self.basedir.joinpath('etc')
+        self.log = self.basedir.joinpath('log')
+        self.data = self.basedir.joinpath('data')
+        self.etc.mkdir(exist_ok=True)
+        self.log.mkdir(exist_ok=True)
+        self.data.mkdir(exist_ok=True)
+        self.port = random.randrange(4000, 4500) if port is None else port
```

Comment:
Might be more robust to do something like this: https://stackoverflow.com/questions/1365265/on-localhost-how-do-i-pick-a-free-port-number

Reply:
It is not clear to me how I would do that - it needs to write a cluster file. Or do you mean I should first run a test program that would chose the port?

Reply:
Done, I wrote a small python function that binds to `0.0.0.0:0` und then gets the port. I believe the OS will assign ports in kind of a round-robin fashion - so even if there are races it seems unlikely enough that the port gets assigned between this function returning and fdbserver starting...